feat add ipmi component

parsa97 opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/alloy/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description
This PR adds a new `prometheus.exporter.ipmi` component to Grafana Alloy for collecting hardware metrics from IPMI-enabled devices.

**Key Features:**
- **Local IPMI Collection**: Monitor the local machine's IPMI interface using FreeIPMI tools
- **Remote IPMI Monitoring**: Monitor network-accessible IPMI devices with authentication support
- **Multiple Target Support**: Configure multiple remote targets with individual credentials
- **Authentication Options**: Supports user/password authentication with configurable driver (LAN_2_0/LAN) and privilege levels (user/admin)
- **Configurable Timeout**: Adjustable timeout for IPMI sensor collection (default: 30 seconds)
- **Module Configuration**: Support for custom collector modules via inline config or external file

**Implementation Details:**
- Component: `internal/component/prometheus/exporter/ipmi/`
- Integration: `internal/static/integrations/ipmi_exporter/`
- Dependencies: Uses `github.com/bougou/go-ipmi` for IPMI operations

#### Which issue(s) this PR fixes

Fixes #731 

#### Notes to the Reviewer

---

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [ ] Config converters updated

parsa97 committed:
(feat) add ipmi component

https://github.com/grafana/alloy/issues/731

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+A timeout of 30 seconds or more is recommended for most hardware.
+
+The `config_file` and `ipmi_config` arguments are mutually exclusive.
+They allow advanced configuration of IPMI collection modules and command overrides.
```
```suggestion
These arguments enable advanced configuration of IPMI collection modules and command overrides.
```

parsa97 replied:
Suggestion applied

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+| `module`    | `string` | IPMI collector module to use.                 |             | no       |
+
+The `target` block defines a remote IPMI device to monitor.
+Multiple `target` blocks can be specified to monitor multiple devices.
```
```suggestion
You can specify multiple `target` blocks to monitor multiple devices.
```

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+On Linux systems, this typically requires:
+- FreeIPMI tools installed
+- Kernel modules loaded (`ipmi_devintf`, `ipmi_si`)
+- Appropriate device permissions (`/dev/ipmi0`)
```
```suggestion
On Linux systems, this typically requires:

- FreeIPMI tools installed
- Kernel modules for `ipmi_devintf` and `ipmi_si` loaded
- Appropriate device permissions for `/dev/ipmi0`
```

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+| `ipmi_collector_info`       | Gauge | Information about the IPMI collector configuration               |
+
+All sensor metrics include labels:
+- `target`: IP address or "localhost" for local collection
```
```suggestion
All sensor metrics include labels:

- `target`: IP address or "localhost" for local collection
```

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+Choose the appropriate IPMI driver based on your hardware:
+
+- **LAN_2_0** (IPMI 2.0): Recommended for modern servers (Dell, HP, Supermicro from ~2010+)
+- **LAN** (IPMI 1.5): For older hardware that doesn't support IPMI 2.0
```
```suggestion
- **LAN** (IPMI 1.5): For legacy hardware that doesn't support IPMI 2.0
```

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+- **Use dedicated monitoring users** with minimal required privileges
+- **Store passwords securely** using `env()` or secrets management
+- **Never commit passwords** to version control
+- **Consider network isolation** for IPMI interfaces (dedicated management network)
```
```suggestion
- **Consider network isolation** for IPMI interfaces on dedicated management networks
```

parsa97 replied:
Done

clayton-cornell commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+
+### Local IPMI requirements
+
+For local IPMI collection, ensure:
```
```suggestion
For local IPMI collection, ensure the following minimum requirements are met:
```

parsa97 replied:
Done

koalalorenzo commented:
@parsa97 do you have any update? :)

parsa97 commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ipmi/
+aliases:
+  - ../prometheus.exporter.ipmi/ # /docs/alloy/latest/reference/components/prometheus.exporter.ipmi/
+description: Learn about prometheus.exporter.ipmi
```
Suggestion applied

parsa97 commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,363 @@
+  stage: general-availability
+  products:
+    - oss
+title: prometheus.exporter.ipmi
```
Suggestion applied

parsa97 committed:
fix: update tablewriter API calls for v1.0.9 compatibility

Update prometheus remotewrite CLI to use the new tablewriter API. The
SetHeader method has been replaced with Header which takes variadic
arguments instead of a string slice.

This fixes the build error:
table.SetHeader undefined (type *tablewriter.Table has no field or method SetHeader)

dehaansa commented on internal/component/prometheus/exporter/ipmi/config.go:
```diff
@@ -0,0 +1,99 @@
+
+// Validate implements syntax.Validator.
+func (a *Arguments) Validate() error {
+	if !a.Local.Enabled && len(a.Targets) == 0 {
```
This should also validate the Driver/Privilege enums. Should it also validate that password is provided if username is configured (is password required or is passwordless auth an option with just user?)

parsa97 commented on docs/sources/reference/components/prometheus/prometheus.exporter.ipmi.md:
```diff
@@ -0,0 +1,354 @@
+
+The `driver` attribute specifies which IPMI protocol version to use:
+
+- `LAN_2_0`: IPMI 2.0 protocol, recommended for most modern hardware
```
I followed the upstream ipmi_exporter convention, LAN and LAN_2_0 are the driver names passed directly to freeipmi. I think we should keep it as it is.
